advanced_counter/apps/advanced_counter/lib/advanced_counter.ex

```diff
@@ -20,13 +20,29 @@ defmodule AdvancedCounter do
   def get("cockroach", counter_id), do: Cockroach.one(Counter, counter_id)
   def get("antidote", counter_id), do: Antidote.one(Counter, counter_id)
 
+  def connection_warmup("cloudsql"), do: Ecto.Adapters.SQL.query(CloudSql, "SELECT 1")
+  def connection_warmup("cockroach"), do: Ecto.Adapters.SQL.query(Cockroach, "SELECT 1")
+  def connection_warmup("antidote"), do: {:ok, nil}
+
   def increment(db, counter_id)
 
   def increment("cloudsql", counter_id),
     do: CloudSql.transaction(postgres_increment_multi(counter_id))
 
-  def increment("cockroach", counter_id),
-    do: Cockroach.transaction(postgres_increment_multi(counter_id))
+  def increment("cockroach", counter_id) do
+    Cockroach.transaction(postgres_increment_multi(counter_id))
+  rescue
+    e in Postgrex.Error ->
+      case e.postgres do
+        %{code: :serialization_failure} ->
+          # Retry the operation per the Cockroach docs
+          # https://www.cockroachlabs.com/docs/v22.1/transactions#client-side-intervention
+          increment("cockroach", counter_id)
```

Comment:
Was wondering whether this could get stuck in an infinite loop. I guess not, but perhaps is it worth limiting the retries to up-to a max of some single digit number of times?

Reply:
Limited retires to 3